add/movie-viewmodel-testing

Mostafa-alsaygh opened the pull request:
# What does this PR do?
make a testing for the MovieDetailsViewModelTest and some refactoring to the viewmodel and tvShowViewModel test

Mostafa-alsaygh committed:
Fix: Rename `getMovieImagesUseCase` to `getMovieImages`

The method `getMovieImagesUseCase` in `ManageMovieDetailsUseCase` has been renamed to `getMovieImages` for consistency and clarity.

This change also includes:
- Updating calls to this method in `MovieDetailsViewModel`.
- Adding unit tests for `MovieDetailsViewModel`.
- Fixing test descriptions in `TvShowDetailsViewModelTest` to use lowercase "when".

Mostafa-alsaygh committed:
Refactor: Simplify tests and improve code readability

- Removed redundant ViewModel recreations in `MovieDetailsViewModelTest`.
- Corrected logic for handling successful rating in `MovieDetailsViewModelTest`.
- Improved formatting and removed unnecessary comments for better readability.
- Made `movieIdOverride` a read-only property in `MovieDetailsViewModel` as it's only set during initialization.

Mostafa-alsaygh committed:
Refactor: MovieDetailsViewModelTest

- Remove unused variable `movieDetails`.
- Add a delay to `whenUiEventsAreSent_thenEffectsAreEmitted` test to ensure the IO dispatcher finishes its work.

Mostafa-alsaygh committed:
Refactor: Replace !! with ? in MovieDetailsViewModelTest

Mostafa-alsaygh committed:
Refactor: Improve MovieDetailsViewModelTest clarity and stability

- **Mock Initialization:** Use `@MockK` for cleaner mock creation.
- **Test Setup:** Introduced a `setupDefaultMocks` method for better organization of common mock behaviors.
- **State Assertion:** Changed `expectMostRecentItem()` to `awaitItem()` in Turbine tests for more predictable behavior when observing `StateFlow`. This ensures the test waits for the first emitted state after an action, rather than potentially missing intermediate states.
- **Error Handling Test:** Added a specific test case for primary initialization failure to ensure the UI correctly reflects an error state.
- **Test Clarity:** Minor wording adjustments in test names for better readability.
- **Lifecycle Management:** Simplified ViewModel lifecycle management in `tearDown`.
- **Redundant `ensureAllEventsConsumed()`:** Removed calls to `ensureAllEventsConsumed()` where they were not strictly necessary, relying on `awaitItem()` for state verification.

london-bot[bot] committed:
Branch was auto-updated.

BassantM commented on presentation/src/test/java/com/london/presentation/feature/details/tvshow/info/MovieDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,391 @@
+@OptIn(ExperimentalCoroutinesApi::class)
+class MovieDetailsViewModelTest {
+
+    @MockK
+    private lateinit var movieDetails: ManageMovieDetailsUseCase
+    @MockK
+    private lateinit var manageRecentMovieWatchedUseCase: ManageRecentMovieWatchedUseCase
+    @MockK
+    private lateinit var manageRecentViewedUseCase: ManageRecentViewedUseCase
+    @MockK
+    private lateinit var ratingUseCase: RatingUseCase
+    @MockK
+    private lateinit var authenticationUseCase: AuthenticationUseCase
+    @MockK(relaxed = true)
```
kindly, follow pattern 
ex: private lateinit var movieDetails: ManageMovieDetailsUseCase = mockk()

Mostafa-alsaygh replied:
this is advance way, why using old one

BassantM replied:
to follow team, untill all cheange it

BassantM commented on presentation/src/test/java/com/london/presentation/feature/details/tvshow/info/MovieDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,391 @@
+    fun setup() {
+        Dispatchers.setMain(mainDispatcher)
+        MockKAnnotations.init(this)
+        movieDetails = mockk()
+        manageRecentMovieWatchedUseCase = mockk()
+        manageRecentViewedUseCase = mockk()
+        ratingUseCase = mockk()
+        authenticationUseCase = mockk()
```
then remove those from setup

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/develop' into add/movie-viewmodel-testing

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt

Mostafa-alsaygh committed:
Test: Enhance unit tests for MovieDetailsViewModel

- Improved clarity and robustness of existing tests.
- Addressed potential flakiness by managing ViewModel lifecycle and using `advanceUntilIdle()` more effectively.
- Ensured proper mocking of dependencies and interactions.
- Added specific tests for edge cases like null movieId and empty image lists.

Fix: Ensure movie video is always selected

- Modified logic to always select the first video if available, preventing potential issues with empty video lists.

london-bot[bot] committed:
Branch was auto-updated.

Abdulrahman-Ragab-01 commented on presentation/src/test/java/com/london/presentation/feature/details/tvshow/info/MovieDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,381 @@
+@OptIn(ExperimentalCoroutinesApi::class)
+class MovieDetailsViewModelTest {
+
+    @MockK
```
Don't use @mockK, make it mock = relaxed in @Before

Mostafa-alsaygh replied:
why?

Mostafa-alsaygh replied:
done

london-bot[bot] committed:
Branch was auto-updated.

Abdulrahman-Ragab-01 commented on presentation/src/test/java/com/london/presentation/feature/details/tvshow/info/MovieDetailsViewModelTest.kt:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    @Test
+    fun `when loadMovieDetails succeeds, all movie details should be populated correctly`() =
```
write all test cases title in this structure (function name should return ......... when .........)

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Test: Enhance unit tests for MovieDetailsViewModel and TvShowDetailsViewModel

- Improved test names to be more descriptive and follow a consistent pattern.
- Refactored mock initialization to be more concise.
- Cleaned up tearDown logic for better resource management.
- Ensured ViewModel scope cancellation for all created ViewModels in tests.
- Adjusted test assertions to be more precise and readable.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
refactor: MovieDetailsScreen

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
refactor: rename MovieDetailsEffect cases for clarity

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
refactor: update MovieDetailsScreen to use callbacks for navigation

Mostafa-alsaygh committed:
Merge branch 'develop' into add/movie-viewmodel-testing

# Conflicts:
#	domain/src/main/java/com/london/domain/usecase/details/movie/ManageMovieDetailsUseCase.kt
#	domain/src/test/kotlin/com/london/domain/usecase/details/movie/GetMovieUseCaseTest.kt

Mostafa-alsaygh committed:
refactor: rename MovieDetailsViewModel test and use GetMovieUseCase

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
refactor: improve MovieDetailsScreen structure and use TextWithIcon

Mostafa-alsaygh committed:
refactor: remove unused fields from MovieDetailsUiState

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

BassantM committed:
Merge remote-tracking branch 'origin/develop' into refactor/movie-details

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsContract.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsEffect.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Merge branch 'refs/heads/develop' into refactor/movie-details

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsScreen.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsUiState.kt
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt

Mostafa-alsaygh committed:
Refactor movie details screen and view model

This commit refactors the MovieDetailsScreen and MovieDetailsViewModel.

Key changes:

-   Simplified the effect handling in `MovieDetailsScreen` by directly calling navigation functions.
-   Removed redundant UI components like separator dots between movie info.
-   Updated `MovieDetailsViewModel` to directly use `MovieDetails`, `List<String>`, and `List<Actor>` instead of a `Triple`.
-   Modified navigation function calls in `mainNavGraph.kt` to match the updated `MovieDetailsScreen` signature.
-   Ensured `MovieGenreUi` is used for genre navigation.
-   Updated `MediaType` usage for reviews navigation.

Mostafa-alsaygh committed:
Merge branch 'refactor/movie-details' into add/movie-viewmodel-testing

# Conflicts:
#	presentation/src/main/java/com/london/presentation/feature/details/movie/MovieDetailsViewModel.kt

Mostafa-alsaygh committed:
fix(details): Rename movie details navigation effects

- Renamed `OnLoginNavigation` to `LoginNavigation`.
- Changed `onReviewsClick` parameter type from `Int` to `MediaType`.
- Changed `onGenreClick` parameter type from `Int` to `MovieGenreUi`.
- Renamed `getMovieImagesUseCase` to `getMovieImages`.
- Removed unnecessary `getMovieId` test.

Mostafa-alsaygh committed:
test(MovieDetailsViewModel): Remove assertion for isLoading

The isLoading state is no longer managed directly within this part of the ViewModel due to changes in how rating updates are handled. This assertion is no longer relevant.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

Mostafa-alsaygh committed:
Remove unnecessary properties from constants.properties

Mostafa-alsaygh committed:
Merge remote-tracking branch 'origin/add/movie-viewmodel-testing' into add/movie-viewmodel-testing

Mostafa-alsaygh committed:
Merge branch 'develop' into add/movie-viewmodel-testing

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.

london-bot[bot] committed:
Branch was auto-updated.
